### PR TITLE
Add lib directory argument to linker

### PR DESCRIPTION
**Pull request by bjfish**
I had an issue trying to link `mysql2` (below). So, I've attempted to handle this argument with this PR. 

One potential issue I'm seeing is that a symlink is being added twice for the same library. I'm not sure if symlinks should be added at all or if any special care needs to happen when a symlink points to the another lib being loaded.
```
$ ls -l /usr/local/Cellar/mysql/5.7.18_1/lib/
total 79936
-r--r--r--   1 bfish  admin   3773144 Jul 16 12:07 libmysqlclient.20.dylib
lrwxr-xr-x   1 bfish  admin        23 Mar 18 02:56 libmysqlclient.dylib -> libmysqlclient.20.dylib
```
```
/Users/bfish/Documents/truffleruby/bin/truffleruby -e Truffle::CExt::Linker.main -- -o mysql2.su client.bc infile.bc mysql2_ext.bc result.bc statement.bc -L/usr/local/Cellar/mysql/5.7.18_1/lib -lmysqlclient -lssl -lcrypto
[ruby] PERFORMANCE this JVM does not have the Graal compiler - performance will be limited - see doc/user/using-graalvm.md
-e:1:in `main': Unknown argument: -L/usr/local/Cellar/mysql/5.7.18_1/lib (RuntimeError)
	from -e:1:in `<main>'
```

### CODE REVIEW

**Comment by eregon** [comment]
The `-L` paths should be passed to Sulong, as the equivalent of their `llvm.libraryPath` option.

**Comment by eregon** [comment]
Therefore they should also be "serialized" in the `.su` and we need to find a way to pass these paths to Sulong (similar to `truffle_load_library()` for `-l`).

**Comment by eregon** [comment]
Actually the lookup method in Sulong is rather simple:
https://github.com/graalvm/sulong/blob/5e6ca402d351b8e38a699cf66da3ce85d630b4f5/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/options/SulongEngineOption.java#L150-L154
https://github.com/graalvm/sulong/blob/5e6ca402d351b8e38a699cf66da3ce85d630b4f5/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/options/SulongEngineOption.java#L202-L213

Therefore it's probably easiest to replicate it here, in the Linker Ruby code.
We need to:
* Collect the search_paths (`-L` options).
* After option parsing, try to resolve libraries which have a relative paths with the `search_paths`, otherwise just add them as-is.

Then `truffle_load_library()` should just work as-is since it will have an absolute path for libmysqlclient.

**Comment by bjfish** [comment]
@eregon I've updated per your suggestions. Please review. The resulting library paths for the given example are: `["/usr/local/Cellar/mysql/5.7.18_1/lib/libmysqlclient.dylib", "libssl.dylib", "libcrypto.dylib"]`

File: src/main/ruby/core/truffle/cext.rb [hunk 1]
```diff
@@ -44,10 +50,33 @@ def self.main(argv = ARGV)
...
+    def self.resolve_libraries(libraries, search_paths)
+      require 'pathname'
+      resolved_libraries = []
+      libraries.each do |lib|
+        if Pathname.new(lib).absolute?
+          resolved_libraries << lib
+        else
+          resolved_libraries << library_in_search_path_or_default(lib, search_paths)
+        end
+      end
```
Comment (eregon) [inline src/main/ruby/core/truffle/cext.rb]: This could be a `map` operation.

File: src/main/ruby/core/truffle/cext.rb [hunk 2]
```diff
@@ -44,10 +50,33 @@ def self.main(argv = ARGV)
...
+      resolved_libraries
+    end
+
+    def self.library_in_search_path_or_default(lib, search_paths)
+      search_paths.each do |search_path|
+        lib_in_path = File.join(search_paths, lib)
```
Comment (eregon) [inline src/main/ruby/core/truffle/cext.rb]: Typo: `search_path`. Maybe name it `path`/`dir` to be visually more distinctive.

File: src/main/ruby/core/truffle/cext.rb [hunk 3]
```diff
@@ -44,10 +50,33 @@ def self.main(argv = ARGV)
...
-
+      libraries = resolve_libraries(libraries, search_paths)
       Truffle::CExt.linker(output, libraries, files)
     end
 
+    def self.resolve_libraries(libraries, search_paths)
+      require 'pathname'
+      resolved_libraries = []
+      libraries.each do |lib|
+        if Pathname.new(lib).absolute?
```
Comment (eregon) [inline src/main/ruby/core/truffle/cext.rb]: It's quite unfortunate there is no `File.absolute?` and loading pathname just for that.
But it's clearer than checking with `start_with?('/')` or `File.absolute_path(lib) == lib`.

File: src/main/ruby/core/truffle/cext.rb (hunk 1 above)
Comment (bjfish) [inline src/main/ruby/core/truffle/cext.rb]: Fixed.

File: src/main/ruby/core/truffle/cext.rb (hunk 2 above)
Comment (bjfish) [inline src/main/ruby/core/truffle/cext.rb]: Fixed.

**Pull request merged**